Add BentBeam analytic solution to 2D linear elasticity

nilsvu opened the pull request:
## Proposed changes

- [x] Depends on #1298 (only the last commit is new)

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsvu committed:
Add BentBeam analytic solution to 2D linear elasticity

mar-celine commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Elasticity/Test_BentBeam.cpp:
```diff
@@ -0,0 +1,79 @@
+namespace {
+
+struct BentBeamProxy : Elasticity::Solutions::BentBeam {
+  using Elasticity::Solutions::BentBeam::BentBeam;
```
Is this line used? (It's unclear to me what this line is doing)

nilsvu replied:
It makes the constructor available to this struct

mar-celine commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Elasticity/Test_BentBeam.cpp:
```diff
@@ -0,0 +1,79 @@
+    "[PointwiseFunctions][Unit][Elasticity]") {
+  const Elasticity::Solutions::BentBeam check_solution{
+      5., 1., 0.5,
+      // Iron: E=100, nu=0.29
```
Are these quantities given for reference? Do the values 79.3651 and 38.7597 correspond to anything?

nilsvu replied:
Sorry, this is unclear at this point because the `IsotropicHomogeneous` constitutive relation is added in #1298. It defines the quantities E and nu in terms of its options K and mu. This PR has a dependence on that.

mar-celine commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.py:
```diff
@@ -0,0 +1,25 @@
+
+
+def source(x):
+    return np.zeros(x.shape)
```
In some AnalyticSolutions directories I have seen the python test functions stored in a TestFunctions.py with a
`# End functions for testing BentBeam.cpp`
at the end.

nilsvu replied:
Please see #1199

mar-celine commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.hpp:
```diff
@@ -0,0 +1,126 @@
+/*!
+ * \brief A state of pure bending of an elastic beam in 2D
+ *
+ * \details This solution describes a 2D slice through an elastic beam of length
```
It would be nice to have the analytic solution here in the form of an equation :+1:

nilsvu replied:
you're right, that's an oversight on my part

mar-celine commented:
@nilsleiffischer you can squash in the fixup :+1:

nilsdeppe commented:
@geoffrey4444 please review this PR or step down as a reviewer.

nilsvu commented:
@geoffrey4444 this is where the story on elasticity continues, please review when you get the chance

kidder commented:
@geoffrey4444 please review this

kidder commented:
@geoffrey4444 ping ping ping

geoffrey4444 commented:
I’ll take a look Monday...I was away this week and unable to get to my usual spectre work

geoffrey4444 commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.hpp:
```diff
@@ -0,0 +1,144 @@
+ * no body-forces \f$f_\mathrm{ext}^j\f$ act on the material, so the
+ * \ref Elasticity equations reduce to \f$\nabla_i T^{ij}=0\f$, but the bending
+ * moment \f$M\f$ generates the stress \f[
+ * T^{xx} = \frac{3M}{2h^3} y \\
```
What is h? Is it the same as H?

nilsvu replied:
yes, just a typo

geoffrey4444 commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.hpp:
```diff
@@ -0,0 +1,144 @@
+ * \f]
+ * we find that this stress is produced by the displacement field
+ * \f{align}
+ * u^x&=-\frac{3M}{2Eh^3}xy
```
Units of M? You call it a moment, and for the units to work it must be a force. Is it just the component of force perpendicular to the undeformed beam? If so, maybe clarify that to help readers unfamiliar with elasticity.

nilsvu replied:
I cite an equation in Thorne/Blandford now, and give a definition

geoffrey4444 commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.hpp:
```diff
@@ -0,0 +1,144 @@
+  struct BendingMoment {
+    using type = double;
+    static constexpr OptionString help{
+        "The bending moment applied to the beam"};
```
Clarify that this is the force causing the displacement (or whatever it should be, including dimensions/units)

geoffrey4444 replied:
Ping ^^^

nilsvu replied:
see below, this is addressed in the help string of the class

geoffrey4444 commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.hpp:
```diff
@@ -0,0 +1,144 @@
+  };
+  struct Height {
+    using type = double;
+    static constexpr OptionString help{"The beam height"};
```
How to units work for elasticity? Are you using geometrized units? Not sure if anything should change, but perhaps clarify units in the help text for your options?

geoffrey4444 replied:
Ping ^^^

nilsvu replied:
This is addressed in the help text of the class, please check that. The individual options' help string must be short. Also I don't think I should add here that "height" is measured in units of length, that seems obvious. That the bending moment is measured in units of force I write in the help string.

geoffrey4444 commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.hpp:
```diff
@@ -0,0 +1,144 @@
+      "An elastic beam subject to a bending moment in 2D"};
+
+  BentBeam() = default;
+  BentBeam(const BentBeam&) noexcept = delete;
```
Just so I understand, please explain to me why you delete some, default others, here?

nilsvu replied:
We don't need the copy constructor since the solution is constructed from options and placed in the cache, so better `delete` it to avoid unintentional copying. We do need the default constructor (for charm and serialization) and the rvalue `&&` functionality for `move`ing the instance around. This is my understanding, @nilsdeppe can you confirm/correct?

nilsvu replied:
confirmed by @nilsdeppe

geoffrey4444 commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.hpp:
```diff
@@ -0,0 +1,144 @@
+  BentBeam& operator=(BentBeam&&) noexcept = default;
+  ~BentBeam() noexcept = default;
+
+  explicit BentBeam(double length, double height, double bending_moment,
```
Why is this marked `explicit`? Again, I only ask because I'm unclear and would like to learn

nilsvu replied:
my bad, no need for the `explicit` here since the constructor takes multiple arguments. If it were only one argument we'd need `explicit` to avoid unintentional conversion from the constructor argument (e.g. `BentBeam solution = 1.`).

geoffrey4444 commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.cpp:
```diff
@@ -0,0 +1,79 @@
+  return {tnsr::I<DataVector, 2>{
+      {{-prefactor * get<0>(x) * get<1>(x),
+        prefactor / 2. *
+            (pow<2>(get<0>(x)) + poisson_ratio * pow<2>(get<1>(x)) -
```
Why not `square()` instead of `pow<2>`? Here and throughout.

nilsvu replied:
sure, changed it

geoffrey4444 commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.cpp:
```diff
@@ -0,0 +1,79 @@
+  const double youngs_modulus = constitutive_relation_.youngs_modulus();
+  const double poisson_ratio = constitutive_relation_.poisson_ratio();
+  const double prefactor =
+      3. * bending_moment_ / (2. * youngs_modulus * pow<3>(height_));
```
Why not `cube()` instead of `pow<3>` here and throughout?

geoffrey4444 commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.py:
```diff
@@ -0,0 +1,25 @@
+
+def displacement(
+        x, length, height, bending_moment, bulk_modulus, shear_modulus):
+    youngs_modulus = 9 * bulk_modulus * shear_modulus / \
```
9 -> 9.0, and similarly everywhere you multiply a hard-coded number by a double, to avoid oddities with e.g. integer vs. double division in some old python versions.

geoffrey4444 commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Elasticity/Test_BentBeam.cpp:
```diff
@@ -0,0 +1,79 @@
+  pypp::check_with_random_values<
+      1, tmpl::list<Tags::Source<Elasticity::Tags::Displacement<2>>>>(
+      &BentBeamProxy::source_variables, solution, "BentBeam", {"source"},
+      {{{-5., 5.}}}, std::make_tuple(), DataVector(5));
```
Is there an analytic solution you can also use in the test? E.g. Thorne and Blandford Eq. (11.46a) is a 1D analytic solution for a bent beam. Maybe there's one for 2D somewhere in the literature?

nilsvu replied:
What do you mean? This _is_ an analytic solution in 2D (i.e. the equivalent to the 1D solution 11.46a in Thorne/Blandford as far as I can tell).

geoffrey4444 replied:
I mean can you check an analytic solution has the correct value, not just check with random values?

nilsvu replied:
Added in another fixup

nilsvu committed:
Add documentation details to IsotropicHomogeneous constitutive relation

nilsvu commented:
@geoffrey4444 Pushed a fixup for you comments.

- I added details about the bending moment, including its definition and mentioning that it is measured in units of force. Also added to the constitutive relation documentation that the elastic moduli are measured in units of stress. I don't think I have to mention that stress is force per area. So when specifying the values in the input file, one makes a choice of units of stress and units of length, and just has to be consistent about it. E.g. one specifies the elastic moduli in Pascals, the lengths in meters and the bending moment in Newtons. Or GPa, km and N.
- I also changed a few constant factors around so that the specified length of the beam is actually the length, and not half the length.

Please let me know when to squash.

nilsvu commented:
@geoffrey4444 could you check the fixup please?

geoffrey4444 reviewed (changes requested):
Please take a look at some comments from my prior review that weren't addressed yet. I also clarified what I meant by requesting a test vs an analytic solution: please test that the solution is correct (say at some particular point) and not just that the random-value checks pass.

nilsvu commented:
@geoffrey4444 the issues are fixed, please see my comments in the threads above from last week. Please let me know when to squash the fixups so that we can finish up this PR

nilsvu commented:
@geoffrey4444 please check this

geoffrey4444 commented:
Please squash. Once Travis is happy, I'm happy to approve

geoffrey4444 reviewed (dismissed):
This is ready for someone else to look at

nilsvu commented:
@gsb76 could you review this again? (Since you did some time in February)

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.hpp:
```diff
@@ -0,0 +1,150 @@
+  BentBeam(double length, double height, double bending_moment,
+           constitutive_relation_type constitutive_relation) noexcept;
+
+  double length() const noexcept { return length_; }
```
why do these (except maybe the constitutive relation) need to be exposed? None of our evolution solutions expose any member variables except the equation of state.

nilsvu replied:
I can't remember why I did that, removed it for now. Kept the `constitutive_relation` accessor because that's probably needed.

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.hpp:
```diff
@@ -0,0 +1,150 @@
+ * no body-forces \f$f_\mathrm{ext}^j\f$ act on the material, so the
+ * \ref Elasticity equations reduce to \f$\nabla_i T^{ij}=0\f$, but the bending
+ * moment \f$M\f$ generates the stress
+ * \f[
```
To avoid clang-format causing problems in the future for the equations could you please put a blank line before the `\f[`, `\f{align}` and after `\f[`, `\f}`?

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.hpp:
```diff
@@ -0,0 +1,150 @@
+/// \endcond
+
+namespace Elasticity {
+namespace Solutions {
```
Is this namespace documented somewhere? I can't find it when going to `Topics->Analytic Solutions` in Doxygen.

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.hpp:
```diff
@@ -0,0 +1,150 @@
+ * \f]
+ * By fixing the rigid-body motions to
+ * \f[
+ * u^x(0,y)=0 \quad \text{and} \quad u^y(\pm \frac{L}{2},0)=0
```
does `\left` and `\right)` not work in Doxygen? The parens here and in the next equation are of the wrong size.

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.hpp:
```diff
@@ -0,0 +1,150 @@
+ * \f]
+ * we find that this stress is produced by the displacement field
+ * \f{align}
+ * u^x&=-\frac{12 M}{EH^3}xy
```
You are missing a new line here.

nilsdeppe reviewed (changes requested):
I'm fine with you squashing.

@gsb76 could you please look at this or let us know you aren't able to review? Thanks! :)

nilsvu committed:
Document Elasticity::Solutions namespace

nilsdeppe commented:
Aside from nothing compiling, :+1:

mar-celine commented:
@nilsleiffischer I've given it a quick pass through again - I didn't see anything I wanted to comment on, I hope I can trust the more recent reviewers in having caught everything :+1:

nilsvu commented:
Also this compiles now, so that's good ;)
